editor/editor_dock_manager.h

```diff
@@ -70,49 +74,51 @@ class EditorDockManager : public Object {
 	};
 
 private:
+	friend class DockContextPopup;
+
+	struct DockInfo {
+		String title;
+		bool open = false;
+		bool enabled = true;
+		bool at_bottom = false;
+		int previous_tab_index = -1;
+		bool previous_at_bottom = false;
+		WindowWrapper *dock_window = nullptr;
+		int dock_slot_index = DOCK_SLOT_LEFT_UL;
+		Ref<Shortcut> shortcut;
+	};
+
 	static EditorDockManager *singleton;
 
 	// To access splits easily by index.
 	Vector<DockSplitContainer *> vsplits;
 	Vector<DockSplitContainer *> hsplits;
 
-	Vector<WindowWrapper *> floating_docks;
-	Vector<Control *> bottom_docks;
+	Vector<WindowWrapper *> dock_windows;
 	TabContainer *dock_slot[DOCK_SLOT_MAX];
+	HashMap<Control *, DockInfo> all_docks;
 	bool docks_visible = true;
 
-	PopupPanel *dock_select_popup = nullptr;
-	Button *dock_float = nullptr;
-	Button *dock_to_bottom = nullptr;
-	Button *dock_tab_move_left = nullptr;
-	Button *dock_tab_move_right = nullptr;
-	Control *dock_select = nullptr;
-	Rect2 dock_select_rect[DOCK_SLOT_MAX];
-	int dock_select_rect_over_idx = -1;
-	int dock_popup_selected_idx = -1;
-	int dock_bottom_selected_idx = -1;
-
-	void _dock_select_popup_theme_changed();
-	void _dock_popup_exit();
-	void _dock_pre_popup(int p_dock_slot);
-	void _dock_move_left();
-	void _dock_move_right();
-	void _dock_select_input(const Ref<InputEvent> &p_input);
-	void _dock_select_draw();
-	void _dock_split_dragged(int p_offset);
+	DockContextPopup *dock_context_popup = nullptr;
+	Control *closed_dock_parent = nullptr;
 
-	void _dock_tab_changed(int p_tab);
-	void _edit_current();
+	void _dock_split_dragged(int p_offset);
+	void _dock_container_gui_input(const Ref<InputEvent> &p_input, TabContainer *p_dock_container);
+	void _bottom_dock_button_gui_input(const Ref<InputEvent> &p_input, Control *p_dock, Button *p_bottom_button);
+	void _dock_container_update_visibility(TabContainer *p_dock_container);
+	void _update_layout();
 
 	void _dock_floating_close_request(WindowWrapper *p_wrapper);
-	void _dock_make_selected_float();
-	void _dock_make_float(Control *p_control, int p_slot_index, bool p_show_window = true);
-	void _restore_floating_dock(const Dictionary &p_dock_dump, Control *p_wrapper, int p_slot_index);
+	Control *_close_floating_window(WindowWrapper *p_wrapper);
+	void _dock_make_float(Control *p_dock, bool p_show_window = true);
+	void _restore_floating_dock(const Dictionary &p_dock_dump, Control *p_wrapper);
```

Comment:
Naming inconsistency.
floating window
floating dock

Reply:
Here's what I'm thinking:

`_dock_floating_close_request` -> `_window_close_request`
`_close_floating_window` -> `_close_window`
`_dock_make_float` -> `_open_dock_in_window`
`_restore_floating_dock` -> `_restore_dock_to_saved_window` and change to only accept dock